apps/web/playwright/limits.e2e.ts

```diff
@@ -0,0 +1,224 @@
+import { expect } from "@playwright/test";
+
+import type { Dayjs } from "@calcom/dayjs";
+import dayjs from "@calcom/dayjs";
+import { entries } from "@calcom/prisma/zod-utils";
+import type { IntervalLimit } from "@calcom/types/Calendar";
+
+import { test } from "./lib/fixtures";
+import { bookTimeSlot } from "./lib/testUtils";
+
+test.describe.configure({ mode: "parallel" });
+test.afterEach(async ({ users }) => {
+  await users.deleteAll();
+});
+
+const EVENT_LENGTH = 30;
+const BOOKING_LIMITS_SINGLE = {
+  day: 2,
+  week: 2,
+  month: 2,
+  year: 2,
+};
+const BOOKING_LIMITS_MULTIPLE = {
+  day: 1,
+  week: 2,
+  month: 3,
+  year: 4,
+};
+
+// prevent tests from crossing year boundaries (if already in Nov, start booking in Jan instead of Dec)
+const firstDayInBookingMonth =
+  dayjs().month() === 10 ? dayjs().add(1, "year").month(0).date(1) : dayjs().add(1, "month").date(1);
+
+// avoid weekly edge cases
+const firstMondayInBookingMonth = firstDayInBookingMonth.day(
+  firstDayInBookingMonth.date() === firstDayInBookingMonth.startOf("week").date() ? 1 : 8
+);
+
+// ensure we land on the same weekday when incrementing month
+const incrementDate = (date: Dayjs, unit: dayjs.ManipulateType) => {
+  if (unit !== "month") return date.add(1, unit);
+  return date.add(1, "month").day(date.day());
+};
+
+// TODO: verify that past bookings count towards limits
+// TODO: consider edge cases, e.g. partial weeks
+
+["booking", "duration"].forEach((limitType) => {
+  test.describe(`${limitType} limits`, () => {
+    entries(BOOKING_LIMITS_SINGLE).forEach(([limitUnit, bookingLimit]) => {
+      const limitValue = limitType === "duration" ? bookingLimit * EVENT_LENGTH : bookingLimit;
+
+      // test one limit at a time
+      test(limitUnit, async ({ page, users }) => {
+        const slug = `${limitType}-limit-${limitUnit}`;
+        const limit = { [`PER_${limitUnit.toUpperCase()}`]: limitValue };
+        const user = await users.create({
+          eventTypes: [{ title: slug, slug, length: EVENT_LENGTH, [`${limitType}Limits`]: limit }],
```

Comment:
`${limitType}Limits` isn't discoverable easily. If we have to change durationLimits to say e.g. durationLimit(hypothetical) then typescript won't be able to discover this I guess. Keeping the code non DRY would help in tests.

Reply:
> typescript won't be able to discover this

Tested this and you are right :eyes:. I've removed all dynamic keys.

Will also add some helpers as you suggested after un-DRYing.